papers/hatra21/hatra21.tex

```diff
@@ -178,31 +194,26 @@ \subsection{Infallible types}
 where some subterms are \emph{flagged} as having type errors, written $\squnder{N}$. Write $\erase(N)$
 for the result of erasing flaggings: $\erase(\squnder{N}) = \erase(N)$.
 
-%% For example the usual
-%% type rules for field access becomes:
-%% \[
-%%   \infer{
-%%     \Gamma \vdash M \Rightarrow M' : T
-%%   }{
-%%     \Gamma \vdash M.\ell \Rightarrow M'.\ell : U
-%%   }
-%%   [
-%%     T = \{ \overline{\ell:U} \} \mbox{ and } (\ell:U) \in (\overline{\ell:U})
-%%   ]
-%% \]
-%% but there is also a rule for unsuccessful field access:  
-%% \[
-%%   \infer{
-%%     \Gamma \vdash M \Rightarrow M' : T
-%%   }{
-%%     \Gamma \vdash M.\ell \Rightarrow \squnder{M'.\ell} : U
-%%   }
-%%   [
-%%     T = \{ \overline{\ell:U} \} \mbox{ implies } \ell \not\in \overline{\ell}
-%%   ]
-%% \]
-%% In this type rule, $U$ is unconstrained.
-
+For example, in Lua, the $\STRING.\FIND$ function expects two strings, and returns the
+offsets for that string:
+\[
+  \STRING.\FIND(\strlit{hello}, \strlit{ell}) \rightarrow (2, 4)
+\]
+and in Luau it has the type:
+\[
+  \STRING.\FIND : (\STRING, \STRING) \rightarrow (\NUMBER?, \NUMBER?)
+\]
+In a conventional type system, there is no judgment for ill-typed terms
+such as $\STRING.\FIND(\strlit{hello}, 37)$ but in an infallible system we flag the error
+and approximate the type, for example:
```

Comment:
I don't follow how this is related to an infallible type system? This is syntactically valid and, from the perspective of most type systems, including Luau, it is valid at the type level as well.

Reply:
Is it, doesn't `BuiltinDefinitions.cpp` give this as having type `(string,string)->(number?, number?)`, and `37` isn't a string!